Refactor apex tests implementation to use apex-node over cli

azlam-abdulsalam opened the pull request:
- refactor(apex-tests): use apex-node and remove dependency on executing cli
- refactor(apex-test): deprecate synchronous mode

Commit:
refactor(apex-tests): use apex-node and remove dependency on executing cli

This change removes the use of sfdx cli to trigger apex tests, instead it will use the recently open
sourced apex-node (which powers the cli) directly. This would help us to provide progress
information in the cli and also helps with further orchestration in an easier manner

BREAKING CHANGE: Synchronous mode no longer supported in sfpowerscripts apex test triggers

Commit:
refactor(apex-tests): use apex-node and remove dependency on executing cli

This change removes the use of sfdx cli to trigger apex tests, instead it will use the recently open
sourced apex-node (which powers the cli) directly. This would help us to provide progress
information in the cli and also helps with further orchestration in an easier manner

BREAKING CHANGE: Synchronous mode no longer supported in sfpowerscripts apex test triggers

Commit:
refactor(apex-test): deprecate synchronous mode

Commit:
refactor(apex-test): deprecate synchronous mode

Commit:
Fix Conflicts

Commit:
Run prettier

Commit:
Merge changes from develop

Commit:
chore(dependencies): update dependencies and versions

Commit:
chore(package-lock): sync package.json and package-lock.json

azlam-abdulsalam committed:
Merge branch 'develop' into refactor/apex-tests

Commit:
Merge branch 'develop' of https://github.com/Accenture/sfpowerscripts into refactor/apex-tests

Commit:
fix(package-lock): rebuild package lock after a incorrect conflict resolution

azlam-abdulsalam committed:
Merge branch 'develop' into refactor/apex-tests

azlam-abdulsalam committed:
chore(prettier): Apply formatting changes

Commit:
feat(apex-test): add elapsed time to polling message

azlam-abdulsalam committed:
Merge branch 'develop' into refactor/apex-tests

azlam-abdulsalam committed:
style(prettier): Apply formatting changes

aly76 commented on packages/core/src/apextest/TriggerApexTests.ts:
```diff
@@ -0,0 +1,334 @@
+                    LoggerLevel.INFO,
+                    this.fileLogger
+                );
+                if (runStatus.Status == `Completed` || runStatus.Status == `Cancelled`) break;
```
The possible Status values are:
- Queued
- Processing
- Aborted
- Completed
- Failed

aly76 replied:
Need to handle Aborted, Completed and Failed, else an infinite loop may occur

azlam-abdulsalam replied:
fixed to use reporter instead of using polling

aly76 commented on packages/core/src/apextest/TriggerApexTests.ts:
```diff
@@ -0,0 +1,334 @@
+    }> {
+        let org = await Org.create({ aliasOrUsername: this.target_org });
+        this.conn = org.getConnection();
+
```
should shutdown gracefully on interrupt/terminate signal by cancelling async job

aly76 commented on packages/core/src/apextest/TriggerApexTests.ts:
```diff
@@ -0,0 +1,334 @@
+                SFPLogger.log(
+                    `Executed Tests ... ${runStatus.ClassesCompleted} of ${
+                        runStatus.ClassesEnqueued
+                    }  Elapsed Time so far: ${Math.floor(Date.now() - startTime) / 1000}`,
```
Need units for time

aly76 commented on packages/core/src/apextest/TriggerApexTests.ts:
```diff
@@ -0,0 +1,334 @@
+            );
+            fs.writeJSONSync(
+                path.join(this.testOptions.outputdir, `test-result-${testRunIdResult.testRunId}-coverage.json`),
+                jsonOutput.coverage.coverage,
```
The result files do not match those created by the CLI.

- `test-result-${testRunId}-codecoverage.json` maps to `TestResult.perClassCoverage`
- `test-result-codecoverage.json` maps to `CLIJsonFormat.coverage.coverage` or an empty array
- `test-result-${testRunId}.json` maps to TestResult
- `test-run-id.txt` maps to TestRunId

azlam-abdulsalam replied:
We will deviate from CLI here, we will provide the following files
 - test-result-${testRunId}-codecoverage.json maps to CLIJsonFormat.coverage.coverage or an empty array
 - test-result-${testRunId}.json maps to TestResult
 - test-result-${testRunId}-junit.xml with junit report

aly76 replied:
The filename being used doesn't match your comment: `test-result-${testRunResult.testRunId}-coverage.json`

aly76 commented on packages/core/tests/sfpcommands/TriggerApexTestSpecifiedTests.test.ts:
```diff
@@ -1,516 +0,0 @@
-import { jest, expect } from '@jest/globals';
```
New test class should be written with appropriate mocks for test service and test result

azlam-abdulsalam replied:
Will be introducing this at later point, as I think it not necessary anymore, as we are using apex-node directly

aly76 commented on packages/core/tests/sfpcommands/TriggerApexTestApexTestInPackage.test.ts:
```diff
@@ -1,539 +0,0 @@
-import { jest, expect } from '@jest/globals';
```
New test class should be written

azlam-abdulsalam replied:
Will be introducing this at later point, as I think it not necessary anymore, as we are using apex-node directly

azlam-abdulsalam replied:
Will be introducing this at later point, as I think it not necessary anymore, as we are using apex-node directly

aly76 commented on packages/core/src/apextest/TriggerApexTests.ts:
```diff
@@ -0,0 +1,334 @@
+
+            //Poll for tests completion
+            let startTime = Date.now();
+            while (true) {
```
Unable to interrupt while loop with signal

Commit:
Merge branch 'develop' of https://github.com/Accenture/sfpowerscripts into refactor/apex-tests

Commit:
refactor(apex-tests): use reporter from apex-node and remove polling

Remove polling and provide progress through apex-node's progrees reporter. Also address some
reporting concerns

BREAKING CHANGE: sfpowerscripts will no longer report test results in the same format as CLI

azlam-abdulsalam committed:
style(prettier): Apply formatting changes

Commit:
fix(deepscan): Fix lint issues

azlam-abdulsalam committed:
Merge branch 'develop' into refactor/apex-tests

aly76 commented on packages/core/src/apextest/TriggerApexTests.ts:
```diff
@@ -0,0 +1,375 @@
+            //Write files
+            fs.writeJSONSync(
+                path.join(this.testOptions.outputdir, `test-result-${testRunResult.testRunId}.json`),
+                JSON.stringify(testResult),
```
`writeJSONSync` expects an object and not a string

azlam-abdulsalam replied:
```suggestion
                testResult,
```

azlam-abdulsalam committed:
fix(apex-test): Utilize object instead of string

azlam-abdulsalam committed:
Merge branch 'develop' into refactor/apex-tests
